src/v/kafka/server/handlers/fetch.h

```diff
@@ -125,6 +151,14 @@ struct op_context {
 
     response_iterator response_end() { return iteration_order.end(); }
 
+    void adopt_fetch_memory_units(fetch_memory_units_t&& units) {
```

Comment:
This doesn't seem to handle the case where `this->shard != units.shard`.

Reply:
It'll likely have an assertion failure in the semaphore's adopt method as the units won't be from the same semaphore instance.

Reply:
So we move semaphore units cross shards?

Reply:
This wouldn't have been an issue initially since `fill_fetch_responses` was always called with the results from a single shard, hence the same semaphore. However, now that we're aggregating units across multiple `fill_fetch_responses` the units will be from different shards and semaphore instances.

Reply:
oohh ok i think I understand what's happening now... sorry the fetch code is abusing my brain cells

Reply:
> So we move semaphore units cross shards?

Yeah, when we read data off of a given shard we consume units from that shards memory sempahore. Then we move those units across the shards with the read data. The destructor for `memory_units_t` makes a cross-shard call to release the units on the proper shard.